WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderPostCardActionsHandler.kt

```diff
@@ -126,22 +131,71 @@ class ReaderPostCardActionsHandler @Inject constructor(
         fromPostDetails: Boolean = false
     ) {
         withContext(bgDispatcher) {
-            when (type) {
-                FOLLOW -> handleFollowClicked(post)
-                SITE_NOTIFICATIONS -> handleSiteNotificationsClicked(post.blogId)
-                SHARE -> handleShareClicked(post)
-                VISIT_SITE -> handleVisitSiteClicked(post)
-                BLOCK_SITE -> handleBlockSiteClicked(post.blogId)
-                LIKE -> handleLikeClicked(post, fromPostDetails)
-                BOOKMARK -> handleBookmarkClicked(post.postId, post.blogId, isBookmarkList, fromPostDetails)
-                REBLOG -> handleReblogClicked(post)
-                COMMENTS -> handleCommentsClicked(post.postId, post.blogId)
-                REPORT_POST -> handleReportPostClicked(post)
-                TOGGLE_SEEN_STATUS -> handleToggleSeenStatusClicked(post, fromPostDetails)
+            var shouldPrefetchSite = false
+            if (type == FOLLOW || type == SITE_NOTIFICATIONS) {
+                val readerBlog = readerBlogTableWrapper.getReaderBlog(post.blogId, post.feedId)
+                if (readerBlog == null) {
+                    shouldPrefetchSite = true
+                }
+            }
+            if (shouldPrefetchSite) {
+                handleActionAfterSiteFetch(post, type, isBookmarkList, fromPostDetails)
+            } else {
+                handleAction(post, type, fromPostDetails, isBookmarkList)
+            }
+        }
+    }
+
+    private suspend fun handleActionAfterSiteFetch(
```

Comment:
💡

It seems the usecase uses Flow, but it always returns a single result. Is the Flow really necessary and does it bring any benefits?

"Collect" is a suspend function, do we need to recursively call "handleAction". We might want to consider taking advantage of the suspend and turn onAction into something like
```
withContext(bgDispatcher) {
            if (type == FOLLOW || type == SITE_NOTIFICATIONS) {
                val readerBlog = readerBlogTableWrapper.getReaderBlog(post.blogId, post.feedId)
                if (readerBlog == null) {
                    val result = fetchSite(post, type, isBookmarkList, fromPostDetails)
                    if (result != FetchSiteState.Success) return@withContext
                }
            } 
            handleAction(post, type, fromPostDetails, isBookmarkList)
        }
```
Wdyt?

Reply:
Thanks for quickly jumping on a call 🙇‍♀️, addressed in 4e4fde2.